File: Algorithms/147.InsertionSortList.cpp
class Solution {
public:
    ListNode *insertionSortList(ListNode *head) {
        // IMPORTANT: Please reset any member data you declared, as
        // the same Solution instance will be reused for each test case.
        if (head == NULL) 
            return head;
        ListNode *p = head->next;
        ListNode *previous = head;
        while (p != NULL) {
            if (p->val >= previous->val) {
                previous = previous->next;
                p = p->next;
                continue;
            }
            previous->next = p->next;
            if (head->val >= p->val) {
                p->next = head;
                head = p;
            } else {
                ListNode *q = head;
                ListNode *q_previous = NULL;
                while (q->val <= p->val) {
                    q_previous = q;
                    q = q->next;
                }
                p->next = q;
                q_previous->next = p;
            }
            p = previous->next;
        }
        return head;
    }
};

File: Algorithms/147.InsertionSortList.py
class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return head
        previous, current = head, head.next;
        while current:
            if current.val >= previous.val:
                previous = previous.next
                current = current.next
                continue
            previous.next = current.next
            if head.val >= current.val:
                current.next = head
                head = current
            else:
                q, q_previous = head, None
                while q.val <= current.val:
                    q_previous = q
                    q = q.next
                current.next = q
                q_previous.next = current
            current = previous.next
        return head
